### ISSUE TITLE
Wait for event implementation question

### ISSUE THREAD

**Issue by michel-m** [chromedp/chromedp#631]
#### What versions are you running?

<pre>
$ go list -m github.com/chromedp/chromedp
github.com/chromedp/chromedp v0.5.3
$ google-chrome --version
83.0.4103.61
$ go version
1.14.3
</pre>

Sorry if I skip the rest of the issue template, it's not so much of an issue than looking for your help to solve my implementation issue.
I'm trying to wait for a whole web page to load before printing it to pdf. The page in question is dynamic, so first I scroll through it to force the various elements to load. Then I would like to implement a wait function until no more of a certain type of network event (such as "load" or "networkIdle") is recorded in the past x seconds.
At the moment, my implementation looks like that (summarized):
```go
network.Enable(),  // though I'm not sure this one is needed
enableLifeCycleEvents(),
chromedp.Navigate(url),
chromedp.WaitReady("body"),
// Then for scrolling the page:
chromedp.EvaluateAsDevTools(`
			var totalHeight = 0;
			var distance = 100;
			var timer = setInterval(() => {
				var scrollHeight = document.body.scrollHeight;
				window.scrollBy(0, distance);
				totalHeight += distance;
				if (totalHeight >= scrollHeight) {
					clearInterval(timer);
				}
			}, 100)
		`, &res),
chromedp.ActionFunc(func(ctx context.Context) error {
			waitForLoadEvent(ctx)
			return nil
		}),
PageToPdf(&buffer),

// With:
func enableLifeCycleEvents() chromedp.ActionFunc {
	return func(ctx context.Context) error {
		err := page.Enable().Do(ctx)
		if err != nil {
			return err
		}
		err = page.SetLifecycleEventsEnabled(true).Do(ctx)
		if err != nil {
			return err
		}
		return nil
	}
}

// and:
func waitForLoadEvent(ctx context.Context) error {
	ch := make(chan struct{})
	cctx, cancel := context.WithCancel(ctx)

	chromedp.ListenTarget(cctx, func(ev interface{}) {
		switch e := ev.(type) {
		case *page.EventLifecycleEvent:
			if e.Name == "load" {
				cancel()
				close(ch)
			}
		}
	})
	select {
	case <-ch:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
```

Of course `waitForLoadEvent` works as intended, which is not as I want it to work. It stops the wait when the first "load" event is recorded, but I would like to make sure that there is no more "load" event in the next x seconds.

Any help would be hugely appreciated, I've spent the last three days on this problem... (Go is new to me.)

Thank you very much in advance!
M.

**Comment by mvdan** [comment]
Is https://github.com/chromedp/chromedp/issues/431 what you want? Or how is that different from the "load" lifecycle events?

**Comment by michel-m** [comment]
Hi @mvdan,
I tried this solution at first, but it wasn't satisfying (the lock was released before the browser had time to scroll down the whole page and trigger the loading of some elements).
I ended up using a blocking loop on DOM elements filtered by class (fortunately this crawler of mine is focused on just one particular website so I could use custom CSS targets).
In any case, thanks for your answer!
